### ISSUE TITLE
U-Net

### ISSUE THREAD

**Issue by yerim0829** [yerim0829/CV-study#2]
import torch
import torch.nn as nn

# ---------------------------------
# Double Convolution Layer
# (Conv → BN → ReLU) × 2
# ---------------------------------
class DoubleConv(nn.Module):
    def __init__(self, in_channels, out_channels):
        super().__init__()
        self.double_conv = nn.Sequential(
            nn.Conv2d(in_channels, out_channels, kernel_size=3, padding=1, bias=False),
            nn.BatchNorm2d(out_channels),
            nn.ReLU(inplace=True),

            nn.Conv2d(out_channels, out_channels, kernel_size=3, padding=1, bias=False),
            nn.BatchNorm2d(out_channels),
            nn.ReLU(inplace=True),
        )

    def forward(self, x):
        return self.double_conv(x)

# ---------------------------------
# U-Net Architecture
# ---------------------------------
class UNet(nn.Module):
    def __init__(self, in_channels=1, num_classes=1, features=[64, 128, 256, 512]):
        super().__init__()

        # Encoder (Down)
        self.downs = nn.ModuleList()
        self.pool = nn.MaxPool2d(kernel_size=2, stride=2)

        for feature in features:
            self.downs.append(DoubleConv(in_channels, feature))
            in_channels = feature

        # Bottleneck
        self.bottleneck = DoubleConv(features[-1], features[-1] * 2)

        # Decoder (Up)
        self.ups = nn.ModuleList()
        self.up_convs = nn.ModuleList()

        for feature in reversed(features):
            self.ups.append(
                nn.ConvTranspose2d(feature * 2, feature, kernel_size=2, stride=2)
            )
            self.up_convs.append(DoubleConv(feature * 2, feature))

        # Final output layer
        self.final_conv = nn.Conv2d(features[0], num_classes, kernel_size=1)

    def forward(self, x):
        skip_connections = []

        # Encoder
        for down in self.downs:
            x = down(x)
            skip_connections.append(x)
            x = self.pool(x)

        # Bottleneck
        x = self.bottleneck(x)

        # Reverse skip list (because decoder goes backward)
        skip_connections = skip_connections[::-1]

        # Decoder
        for idx in range(len(self.ups)):
            x = self.ups[idx](x)

            skip_connection = skip_connections[idx]

            # Handle odd size issues
            if x.shape != skip_connection.shape:
                x = nn.functional.interpolate(x, size=skip_connection.shape[2:])

            x = torch.cat([skip_connection, x], dim=1)
            x = self.up_convs[idx](x)

        return self.final_conv(x)
